### PR TITLE
Add experimental remote MDX support to esbuild

### PR DESCRIPTION
**Pull request by wooorm**
This adds support for loading MD(X) files over the wire (`https?://`). This might be interesting for folks who have files in other repos for example. Or that want to include some readme from GitHub and such.
    
Note though that this is a) unsafe: it evals remote code, and b) experimental: Please test it out, see whether this is a good idea or not, and leave feedback!

*[Review w/o whitespace](https://github.com/wooorm/xdm/pull/66/files?w=1)*

*[See docs](https://github.com/wooorm/xdm/blob/remote-mdx/readme.md#importing-md-and-mdx-files-from-the-web-in-esbuild)*

### CODE REVIEW

File: lib/integration/esbuild.js [hunk 1]
```diff
@@ -2,127 +2,206 @@
...
- * @param {ProcessorOptions} [options]
+ * @param {Options} [options]
  * @return {Plugin}
  */
-export function esbuild(options) {
+export function esbuild(options = {}) {
+  const {allowDangerousRemoteMdx, ...rest} = options
   const name = 'esbuild-xdm'
-  const {extnames, process} = createFormatAwareProcessors(options)
+  const remoteNs = name + '-remote'
```
Comment (ChristianMurphy) [inline lib/integration/esbuild.js:38]:
```suggestion
  const remoteNamespace = name + '-remote'
```

File: lib/integration/esbuild.js [hunk 2]
```diff
@@ -2,127 +2,206 @@
...
-        length = 1
-
-        if (end.line != null && end.column != null && end.offset != null) {
-          length = end.offset - start.offset
-        }
+    build.onLoad({filter: /.*/, namespace: remoteNs}, onloadremote)
+    build.onLoad({filter}, onload)
+
+    /** @param {OnResolveArgs} args  */
+    function resolveRemoteInLocal(args) {
```
Comment (ChristianMurphy) [inline lib/integration/esbuild.js:72]: would:

```suggestion
    function resolveRemoteInLocal({path}) {
```

work here?

File: lib/integration/esbuild.js [hunk 3]
```diff
@@ -2,127 +2,206 @@
...
+      return {path: args.path, namespace: remoteNs}
+    }
+
+    // Intercept all import paths inside downloaded files and resolve them against
+    // the original URL. All of these
+    // files will be in the "http-url" namespace. Make sure to keep
+    // the newly resolved URL in the "http-url" namespace so imports
+    // inside it will also be resolved as URLs recursively.
+    /** @param {OnResolveArgs} args  */
+    function resolveInRemote(args) {
```
Comment (ChristianMurphy) [inline lib/integration/esbuild.js:82]:
```suggestion
    function resolveInRemote({path, importer}) {
```

File: lib/integration/esbuild.js [hunk 4]
```diff
@@ -2,127 +2,206 @@
...
-          length: Math.min(length, lineEnd),
-          lineText: doc.slice(lineStart, lineEnd)
-        },
-        detail: message
-      })
+    /**
+     * @param {OnLoadArgs} data
+     * @returns {Promise<OnLoadResult>}
+     */
+    async function onloadremote(data) {
```
Comment (ChristianMurphy) [inline lib/integration/esbuild.js:93]:
```suggestion
    async function onloadremote({path}) {
```

**Comment by ChristianMurphy** [comment]
Potentially outside the scope of this PR.
Loading from URLs opens more potential for untrusted code to be included, which you have already noted in the docs.

This PR is setup nicely to support trusted files coming from a remote server.
It could be worth seeing if some containerization could be done to limit what markdown and mdx files can access or modify.
Some related projects which can help with sandboxing/containerizing code:
* https://github.com/asvd/jailed#readme
* https://github.com/endojs/endo/tree/master/packages/ses#readme
* https://github.com/patriksimek/vm2#readme

File: lib/integration/esbuild.js (hunk 4 above)
Comment (wooorm) [reply lib/integration/esbuild.js:93]: Not a fan of destructuring *in* params,
- The error is hard to read if a weird value is given: `Cannot destructure property 'path' of 'object null' as it is null`
- hard to figure out / log what else was given
- I’m fine with defaults *in* params, or destructuring after (e.g., [here](https://github.com/wooorm/xdm/pull/66/files#diff-59d6233a7bb3ebea2d25c8defe17f770568246e1d793ee59b9caa4368dc647abR36)), but no need to store stuff in variables for these three imo

**Comment by wooorm** [comment]
Last I looked, for the Security section, I found that vm2 might be the best of those. Not an expert at all though. (note: esbuild will most often be used to create browser bundles though)

I improved the note in the `allowDangerousRemoteMdx` section to link to Security.
Do you have recommendations for the Security section?

**Comment by wooorm** [comment]
Released! https://github.com/wooorm/xdm/releases/tag/1.12.0

**Pull request merged**